### PR TITLE
Light weight source linker

### PR DESCRIPTION
**Pull request by itsgreggreg**
When I was going through the code for moving comment parsing to the lexer I noticed this code could be cleaned up a little.

### CODE REVIEW

**Comment by itsgreggreg** [comment]
The diff's a little janky here, might be easier to look at the source directly https://github.com/gleam-lang/gleam/blob/f150e1623056ebec7951e4f79ffe3671e789fd1f/src/docs/source_links.rs

**Comment by itsgreggreg** [comment]
The diff here is kinda janky. The gist is the SrcLinker only needs to know about line indexes and what the url parts look like. So it calculates that up front if it can and doesn't carry anything else with it. No more `Box`, `dyn`, `Trait` `codespan` or `clone`. about .5 the code size.

File: src/docs/source_links.rs [hunk 1]
```diff
@@ -3,156 +3,87 @@ use crate::{
...
-    let end_line = codespan_file
-        .line_index((), location.end)
-        .unwrap_or_default()
-        + 1;
-
-    (start_line, end_line)
+// get all the line start byte offsets for a str
+fn line_starts(src: &str) -> Vec<usize> {
+    std::iter::once(0)
+        .chain(src.match_indices('\n').map(|(i, _)| i + 1))
```
Comment (lpil) [inline src/docs/source_links.rs:71]: Does this work when non `\n` line endings are used?

**Review by lpil — COMMENTED** [review]
Hey clever. I like the format string bit, much simpler.

File: src/docs/source_links.rs (hunk 1 above)
Comment (itsgreggreg) [reply src/docs/source_links.rs:71]: Oh good question, if the line endings are `\n`(posix) or `\r\n`(win) it should be fine. Will not if they are just `\r`(vintage).

It's copy-pasta from codespan: https://github.com/brendanzab/codespan/blob/master/codespan-reporting/src/files.rs#L251

Comment (lpil) [reply src/docs/source_links.rs:71]: That's probably ok I think? What do you reckon?

Comment (itsgreggreg) [reply src/docs/source_links.rs:71]: I think it's probably fine, kinda deferring judgement to codespan.

Comment (lpil) [reply src/docs/source_links.rs:71]: Let's go then :)

**Review by lpil — APPROVED** [review]

**Pull request merged**